### PR TITLE
[DPO] add KTO loss

### LINKED ISSUE #1072: Implement KTO
https://github.com/ContextualAI/HALOs/blob/main/assets/report.pdf

### PR DESCRIPTION
**Pull request by kashif**
add Tahneman-Tversky optimization (KTO) loss function from https://github.com/ContextualAI/HALOs]

fixes #1072

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -465,8 +465,16 @@ def dpo_loss(
...
             # eqn (17) of the paper where beta is the regularization parameter for the IPO loss, denoted by tau in the paper.
             losses = (logits - 1 / (2 * self.beta)) ** 2
+        elif self.loss_type == "kto":
+            # eqn (7) of the HALOs paper
+            chosen_KL = (policy_chosen_logps - reference_chosen_logps).mean().clamp(min=0)
+            rejected_KL = (policy_rejected_logps - reference_rejected_logps).mean().clamp(min=0)
+
+            chosen_logratios = (policy_chosen_logps - reference_chosen_logps)
+            rejected_logratios = (policy_rejected_logps - reference_rejected_logps)
+            losses = torch.cat((1 - F.sigmoid(self.beta * (chosen_logratios - rejected_KL)), 1 - F.sigmoid(self.beta * (chosen_KL - rejected_logratios))), 0)
```
Comment (lewtun) [inline trl/trainer/dpo_trainer.py:475]: Let's add this since I was confused at first:
```suggestion
            # As described in the KTO report, the KL term for chosen (rejected) is estimated using the rejected (chosen) half.
            losses = torch.cat((1 - F.sigmoid(self.beta * (chosen_logratios - rejected_KL)), 1 - F.sigmoid(self.beta * (chosen_KL - rejected_logratios))), 0)
```

**Review by lewtun — APPROVED** [review]
Super fast addition @kashif ! I've checked the equations and it looks good to me 🔥

File: docs/source/dpo_trainer.mdx [hunk 2]
```diff
@@ -80,14 +80,16 @@ dpo_trainer.train()
...
 
 The [IPO](https://arxiv.org/abs/2310.12036) authors provide a deeper theoretical understanding of the DPO algorithms and identify an issue with overfitting and propose an alternative loss which can be used via the `loss_type="ipo"` argument to the trainer.
 
-The [cDPO](https://ericmitchell.ai/cdpo.pdf) is a tweak on the DPO loss where we assume that the preference labels are noisy with some probability that can be passed to the `DPOTrainer` via `label_smoothing` argument (between 0 and 0.5) and then a conservative DPO loss is used.
+The [cDPO](https://ericmitchell.ai/cdpo.pdf) is a tweak on the DPO loss where we assume that the preference labels are noisy with some probability that can be passed to the `DPOTrainer` via `label_smoothing` argument (between 0 and 0.5) and then a conservative DPO loss is used. Use the `loss_type="cdpo"` argument to the trainer to use it.
+
+The [KTO](https://github.com/ContextualAI/HALOs/blob/main/assets/report.pdf) loss is derived to directly maximize the utility of LLM generations instead of the log-likelihood of prefereces. Use the `loss_type="kto"` argument to the trainer to utilize this loss.
```
Comment (vwxyzjn) [inline docs/source/dpo_trainer.mdx:92]: Maybe also mention for KTO, we do not need the preference dataset and just need to say if a sample is desirable vs undesirable.

**Review by vwxyzjn — APPROVED** [review]
Nice work @kashif! This is blazing fast. Maybe in a follow-up PR you can add a test example using anthropic's hh?
